packages/freezed/lib/src/freezed_generator.dart

```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
   }
 
   List<Parameter> _commonParametersBetweenAllConstructors(
-    List<ConstructorDetails> constructorsNeedsGeneration,
-  ) {
+    List<ConstructorDetails> constructorsNeedsGeneration, {
+    required bool allowCommonSuperType,
+  }) {
     return constructorsNeedsGeneration.first.parameters.allParameters
         .map((parameter) {
-          var hasAnyFinalProperty = false;
-          for (final constructor in constructorsNeedsGeneration) {
-            final matchingParameter =
-                constructor.parameters.allParameters.firstWhereOrNull((p) {
-              return p.name == parameter.name && p.type == parameter.type;
-            });
+          var anyMatchingPropertyIsFinal = false;
+          var isCommonWithDifferentNullability = false;
+
+          DartType? commonType;
+
+          for (final constructor in constructorsNeedsGeneration.skip(1)) {
+            final matchingParameter = constructor.parameters.allParameters
+                .firstWhereOrNull((p) => p.name == parameter.name);
 
             if (matchingParameter == null) return null;
-            if (matchingParameter.isFinal) hasAnyFinalProperty = true;
-          }
 
-          if (hasAnyFinalProperty) {
-            return parameter.copyWith(isFinal: true);
+            final commonTypeWithMatch =
+                parameter.parameterElement!.library!.typeSystem.leastUpperBound(
+              parameter.parameterElement!.type,
+              matchingParameter.parameterElement!.type,
+            );
+
+            if (commonTypeWithMatch.isDartCoreObject) return null;
+
+            if (!allowCommonSuperType &&
+                commonTypeWithMatch.getDisplayString(withNullability: false) !=
+                    parameter.type?.replaceAll(r'?$', '')) return null;
```

Comment:
Why not `parameter.parameterElement.type.getDisplayString(withNullability: false)`, instead of manually removing nullability? Additionally shouldn't the `!allowCommonSuperType` check occur for determining the `commonType` as well? Specifically in the copyWith below here you might want to only use `commonType` if `allowCommonSuperType` is true.